### ISSUE TITLE
cbp mode cannot break out of the loop to perform key feature initialization operations

### ISSUE THREAD

**Issue by wrench1997** [shibhansh/loss-of-plasticity#5]
# Issue with Python PPO cbp Model in Training Loop

## Description
I'm working with a Python project using a reinforcement learning setup, specifically a PPO (Proximal Policy Optimization) algorithm configured for 'cbp' mode. Here is the relevant section of the launch configuration in my `launch.json`:

```markdown

```json
{
    "name": "Python PPO cbp",
    "type": "debugpy",
    "request": "launch",
    "program": "run_ppo.py",
    "console": "integratedTerminal",
    "args": [
        "-c",
        "cfg/ant/cbp.yml"
    ],
    "justMyCode": false
}
```

## Problem
During the training loop, the code consistently hits a `continue` statement, which I suspect is causing some issues with the training process. Here is the snippet of the problematic part:

```python
"""
Find the no. of features to replace
"""
eligible_feature_indices = torch.where(self.ages[i] > self.maturity_threshold)[0]
if eligible_feature_indices.shape[0] == 0:
    continue
```

It seems that the condition `eligible_feature_indices.shape[0] == 0` is always true, which makes the training loop skip important steps.Given that this issue persists, I am concerned that the 'cbp' mode does not exhibit any differences from the standard PPO configuration.

This is the configuration of cbp mode:

```yaml 
env_name: 'Ant-v3'
act_type: 'ReLU'
n_steps: 1e8
g: 0.99
lm: 0.95
u_epi_ups: 0
bs: 2048
n_itrs: 10
n_slices: 16
lr: 0.0001
h_dim: [256, 256]
u_adv_scl: 1
clip_eps: 0.2
ob_norm: 0
rew_norm: 0
rr: 1e-4
mt: 10000
util_type_val: 'contribution'
util_type_pol: 'contribution'
beta_1: 0.99
beta_2: 0.99
decay_rate: 0
wd: 1e-4
dir: 'data/ant/cbp/'
label: 'CBP (instantaneous)'
```

**_mt == 10000 ?_**

## Questions
1. Why might the `eligible_feature_indices` always be empty?
2. How can I adjust the `maturity_threshold` or any other parameters to ensure that the condition allows the progression of the training loop?
3. Is there something I'm missing in the configuration that could be causing this issue?

Any insights or suggestions would help. Thank you in advance!

**Comment by shibhansh** [comment]
Thank you for raising the concern. Before I take a deeper look at it, can you please tell me how long you ran the experiment? For the current configuration, i.e. mt=10000, bs: 2048, n_itrs: 10, n_slices: 16. The first reinitialization operation will happen after 128,000 time-steps (mt * bs /(n_itrs * n_slices)).

**Comment by wrench1997** [comment]
> Thank you for raising the concern. Before I take a deeper look at it, can you please tell me how long you ran the experiment? For the current configuration, i.e. mt=10000, bs: 2048, n_itrs: 10, n_slices: 16. The first reinitialization operation will happen after 128,000 time-steps (mt * bs /(n_itrs * n_slices)).

Hi there,

After careful review and a significant waiting period, I managed to hit breakpoints successfully, suggesting the code is functioning as expected. Despite this, I'm experiencing prolonged execution times with the current configuration.

**Comment by shibhansh** [comment]
I'm glad to hear that the issue is resolved. One thing you can try to speed up the code is to use the new [CBP-layer implementation](https://github.com/shibhansh/loss-of-plasticity/blob/main/lop/algos/cbp_linear.py).
